src/public.rs

```diff
@@ -0,0 +1,283 @@
+        self.0.as_ref()
+    }
+}
+
+impl From<PublicKey> for [u8; PublicKey::LEN] {
+    fn from(salt: PublicKey) -> [u8; PublicKey::LEN] {
+        salt.0.into()
+    }
+}
+
+impl fmt::LowerHex for PublicKey {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        self.0.fmt(f)
+    }
+}
+
+impl fmt::UpperHex for PublicKey {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        self.0.fmt(f)
+    }
+}
+
+impl fmt::Debug for PublicKey {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        self.0.fmt(f)
+    }
+}
+
+impl fmt::Display for PublicKey {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        self.0.fmt(f)
+    }
+}
+
+#[cfg(feature = "std")]
+mod use_std {
+    use super::*;
+    use crate::{Error, SecretKey, Signature};
+
+    use secp256k1::{
+        recovery::{RecoverableSignature, RecoveryId},
+        {Error as Secp256k1Error, Message, PublicKey as Secp256k1PublicKey, Secp256k1},
+    };
+
+    use core::borrow::Borrow;
+    use core::str;
+
+    const UNCOMPRESSED_PUBLIC_KEY_SIZE: usize = 65;
+
+    impl PublicKey {
+        /// Convert an uncompressed public key representation into self.
+        ///
+        /// # Safety
+        ///
+        /// Will not check elliptic-curve correctness.
+        pub unsafe fn from_uncompressed_unchecked(
+            pk: [u8; UNCOMPRESSED_PUBLIC_KEY_SIZE],
+        ) -> PublicKey {
+            debug_assert_eq!(
+                UNCOMPRESSED_PUBLIC_KEY_SIZE,
+                secp256k1::constants::UNCOMPRESSED_PUBLIC_KEY_SIZE
+            );
+
+            // Ignore the first byte of the compressed flag
+            let pk = &pk[1..];
+
+            // Safety: compile-time assertion of size correctness
+            Self::from_slice_unchecked(pk)
+        }
+
+        /// Check if the provided slice represents a public key that is in the
+        /// curve.
+        ///
+        /// # Safety
+        ///
+        /// This function extends the unsafety of
+        /// [`PublicKey::as_ref_unchecked`].
+        pub unsafe fn is_slice_in_curve_unchecked(slice: &[u8]) -> bool {
+            use secp256k1::ffi::{self, CPtr};
+
+            let public = Self::as_ref_unchecked(slice);
+
+            let mut public_with_flag = [0u8; UNCOMPRESSED_PUBLIC_KEY_SIZE];
+
+            (&mut public_with_flag[1..]).copy_from_slice(public.as_ref());
+
+            // Safety: FFI call
+            let curve = ffi::secp256k1_ec_pubkey_parse(
+                ffi::secp256k1_context_no_precomp,
+                &mut ffi::PublicKey::new(),
+                public_with_flag.as_c_ptr(),
+                UNCOMPRESSED_PUBLIC_KEY_SIZE,
+            );
+
+            curve == 1
+        }
+
+        /// Check if the secret key representation is in the curve.
+        pub fn is_in_curve(&self) -> bool {
+            // Safety: struct is guaranteed to reference itself with correct len
+            unsafe { Self::is_slice_in_curve_unchecked(self.as_ref()) }
```

Comment:
Do we really need this unsafe conversion? Since `self.0` already contains a valid Bytes64, we should be able to preserve the size instead of converting to a slice right?

Reply:
The `is_in_curve` will take an instance of `&self`, while `is_slice_in_curve_unchecked` will just take a slice. They will be used in different places, but `is_slice_in_curve_unchecked` is unsafe because the len of the provided slice won't be checked.